fitparse/encoder.py

```diff
@@ -0,0 +1,440 @@
+import os
+import re
+import struct
+
+import six
+
+from fitparse import FitFileDataProcessor, profile, utils
+from .records import Crc, DataMessage, DefinitionMessage, MessageHeader, FieldData, FieldDefinition
+from .utils import fileish_open, FitParseError
+
+
+class FitFileEncoder(object):
+    def __init__(self, fileish,
+                 protocol_version=None, profile_version=None,
+                 data_processor=None):
+        """
+        Create FIT encoder.
+
+        :param fileish: file-ish object,
+        :param protocol_version: protocol version, change to 2.0 if you use developer fields.
+        :param profile_version: profile version.
+        :param data_processor: custom data processor.
+        """
+        self.protocol_version = 1.0 if protocol_version is None else float(protocol_version)
+        self.profile_version = 20.33 if profile_version is None else float(profile_version)
+
+        self._processor = data_processor or FitFileDataProcessor()
+        self._file = fileish_open(fileish, 'wb')
+        self._byte_start = 0
+        self._bytes_written = 0
+        self._compressed_ts_accumulator = 0
+        self._local_mesgs = {}
+        self.data_size = 0
+        self.completed = False
+        self._crc = Crc()
+
+        self._write_file_header_place()
+
+    def __del__(self):
+        self.close()
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *_):
+        self.close()
+
+    def close(self):
+        self.finish()
+        if hasattr(self, "_file") and self._file and hasattr(self._file, "close"):
+            self._file.close()
+            self._file = None
+
+    ##########
+    # Private low-level utility methods for writing of fit file
+
+    def _write(self, data):
+        if not data:
+            return
+        self._file.write(data)
+        self._bytes_written += len(data)
+        self._crc.update(data)
+
+    def _write_struct(self, data, fmt, endian='<'):
+        if fmt.startswith('<') or fmt.startswith('>'):
```

Comment:
In general, allowing multiple ways to specify a parameter can get confusing (how do they interact?). In this case I would either raise an error if the endianness is specified in both parameters or just not allow it in the `fmt` at all since (as far as I can tell) it's only used for the CRC.

Reply:
IMO it's a design flaw in the fitparse base, already. E.g. see `test.py` func `generate_fitfile()` - there is copy&paste of the `struct.pack` spec for the FIT header and Crc. I've tried to fix that for Crc at least. But I can copy&paste the endianness for the Crc, too.